modules/checklist/steps/step-base.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+namespace Elementor\Modules\Checklist\Steps;
+
+use Elementor\Core\Isolation\Wordpress_Adapter;
+use Elementor\Modules\Checklist\Module;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+abstract class Step_Base {
+	const MARKED_AS_DONE_KEY = 'is_marked_done';
+	const COMPLETED_KEY = 'is_completed';
+
+	/**
+	 * @var string
+	 * This is the key to be set to true if the step can be completed, and still be considered completed even if the user later did something to the should have it marked as not completed
+	 */
+	const IS_ONE_COMPLETION_SUFFICIENT = 'is_completed_once';
+
+	protected $step_data;
+	protected $user_progress;
+	protected Wordpress_Adapter $wordpress_adapter;
+
+	/** @var Module $module */
+	protected $module;
+
+	/**
+	 * Returns a steps current completion status
+	 *
+	 * @return bool
+	 */
+	abstract protected function get_completion_absolute_status();
+
+	/**
+	 * @return string
+	 */
+	abstract public function get_id() : string;
+
+	/**
+	 * @return array
+	 */
+	abstract public function get_config() : array;
+
+	/**
+	 * Step_Base constructor.
+	 *
+	 * @param array $step_data
+	 * @param Module $module
+
+	 * @return void
+	 */
+	public function __construct( $module, ?Wordpress_Adapter $wordpress_adapter = null ) {
+		$this->step_data = $this->get_config();
+		$this->module = $module;
+		$this->wordpress_adapter = $wordpress_adapter ?? new Wordpress_Adapter();
+		$this->user_progress = $module->get_step_progress( $this->get_id() ) ?? $this->get_step_initial_progress();
+	}
+
+	/**
+	 * Marking a step as done based on user's desire
+	 *
+	 * @return void
+	 */
+	public function mark_as_done() : void {
```

Comment:
should we return true/false when succeed or not?

Reply:
No real requirement for it, depends how polite we're feeling like
WDYT?